src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -763,54 +847,229 @@ cbuf_unmap_at(spdid_t s_spd, unsigned int cbid, spdid_t d_spd, vaddr_t d_addr)
+			struct cbuf_meta *cm = m->m;		/* Use this after loop */
+
+			do {
+				printc("[%d:%x] ", m->spdid, m->addr);
+				m = FIRST_LIST(m, next, prev);
+			} while (m != &cbi->owner);
+
+			printc("\n\tnfo: [ex %d|ow %d|t %d|in %d|re %d] page pointer %x",
+									CBUF_EXACTSZ(cm)      >> 11,
+									CBUF_OWNER(cm)        >> 10,
+									CBUF_TMEM(cm)         >>  9,
+									CBUF_INCONSISTENT(cm) >>  8,
+									CBUF_RELINQ(cm)       >>  7,
+									CBUF_PTR(cm));
+			
+			printc("\n");
+		}
+
+		current = FIRST_LIST(current, next, prev);
+	} while (current != cbi_head);
+
+	printc("Done iterating over global cbuf list\n");
+}
+
+/* This is internal so the real core of the function doesn't take a lock */
+static vaddr_t
+__cbuf_fork_spd(spdid_t o_spd, spdid_t f_spd, int cinfo_cbid)
+{
+	struct cbuf_comp_info *src, *dst;
+	struct cbuf_info *current;
+	int i;
+	vaddr_t ret = (vaddr_t) NULL;		/* address of cinfo page */
+
+	printc("starting valloc_fork_spd from %d to %d\n", o_spd, f_spd);
+	int r = valloc_fork_spd(cos_spd_id(), o_spd, f_spd);
+	printc("finished valloc_fork_spd with ret %d\n", r);
+	
+	src = cbuf_comp_info_get(o_spd);
+	dst = cbuf_comp_info_get(f_spd);
+	__get_nfo(src);
+
+	/* Should create a new shared page between cbuf_mgr and dst */
+	dst->csp = NULL;
+	//dst->dest_csp = src->dest_csp;
+	
+	current = cbi_head;
+	vaddr_t r_addr;
+	do {
+		// The cbuf_info list we made is actually crap because it copies the pointers so if the structure is updated it is no longer valid (how to fix???)
+		// instead, re-get the cbi from the cbuf_id which we assume is constant
+		struct cbuf_info *cbi = cmap_lookup(&cbufs, current->cbid);
+		struct cbuf_maps *m = &cbi->owner;
+	
+		if (m->spdid == o_spd) {
+			/* This is for if O is the owner */
+			/* This just universally forks everything to a new cbuf. */
+			__cbuf_fork_cbuf(o_spd, cbi->cbid, f_spd, 0);
+		}
+		else {
+			do {
+				/* This is if O isn't the owner but has the cbuf mapped in. */
+				if (m->spdid == o_spd) {
+					r_addr = __cbuf_fork_cbuf(o_spd, cbi->cbid, f_spd, cinfo_cbid == cbi->cbid);
+					if (cinfo_cbid == cbi->cbid) ret = r_addr;			// Remember to make this assignment less stupid later
+				}
+
+				m = FIRST_LIST(m, next, prev);
+			} while (m != &cbi->owner);
+		}
+
+		current = FIRST_LIST(current, next, prev);
+	} while (current != cbi_head);
+
+	__get_nfo(dst);
+
+done:
+	return ret;
+}
+
+vaddr_t
+cbuf_fork_spd(spdid_t spd, spdid_t s_spd, spdid_t d_spd, int cinfo_cbid)
+{
+	vaddr_t ret = (vaddr_t) NULL;
 	printl("cbuf_fork_spd\n");
 
 	CBUF_TAKE();
-	s_cci = cbuf_comp_info_get(s_spd);
-	if (unlikely(!s_cci)) goto done;
-	d_cci = cbuf_comp_info_get(d_spd);
-	/* FIXME: This should be making copies to avoid sharing */
-	__cbuf_copy_cci(s_cci, d_cci);
-
+	ret = __cbuf_fork_spd(s_spd, d_spd, cinfo_cbid);
 done:
 	CBUF_RELEASE();
 	return ret;
 }
 
-void cos_fix_spdid_metadata(spdid_t o_spd, spdid_t f_spd)
+void 
+cos_fix_spdid_metadata(spdid_t o_spd, spdid_t f_spd)
```

Comment:
I might be wrong about this but I think this function is required to exist because the fault handler calls it but then does not have to actually do anything. I will verify that.